src/sharing/ShareWrapper.js

```diff
@@ -0,0 +1,252 @@
+  dispatch: Dispatch,
+  auth: Auth,
+  ownUserId: UserId,
+|}>;
+
+type State = $ReadOnly<{|
+  message: string,
+  sending: boolean,
+  files: File[],
+|}>;
+
+/**
+ * Wraps Around different sharing screens,
+ * for minimal duplication of code.
+ */
+class ShareWrapper extends React.Component<Props, State> {
+  static contextType = TranslationContext;
+  context: GetText;
+
+  state = (() => {
+    const { sharedData } = this.props;
+    return {
+      message: sharedData.type === 'text' ? sharedData.sharedText : '',
+      sending: false,
+      files: sharedData.type === 'text' ? [] : sharedData.files,
+    };
+  })();
+
+  handleMessageChange = (message: string) => {
+    this.setState({ message });
+  };
+
+  setSending = () => {
+    this.setState({ sending: true });
+  };
+
+  /**
+   * Send received shared data as a message.
+   *
+   * Sends the shared data received from a 3rd party app and possibly modified
+   * by the user, to the server, as a message.
+   */
+  handleSend = async () => {
+    const _ = this.context;
+    const { auth, sendTo, sharedData } = this.props;
+    let messageToSend = this.state.message;
+
+    this.setSending();
+    showToast(_('Sending Message...'));
+    if (sharedData.type === 'file') {
+      const { files } = this.state;
+      for (let i = 0; i < files.length; i++) {
+        const response = await api.uploadFile(auth, files[i].url, files[i].name);
+        messageToSend += `\n[${files[i].name}](${response.uri})\n`;
+      }
+    }
+    const messageData =
+      sendTo.type === 'pm'
+        ? {
+            content: messageToSend,
+            type: 'private',
+            to: JSON.stringify(sendTo.selectedRecipients),
+          }
+        : {
+            content: messageToSend,
+            type: 'stream',
+            subject: sendTo.topic,
+            to: sendTo.stream,
+          };
+
+    try {
+      await api.sendMessage(auth, messageData);
+    } catch (err) {
+      showToast(_('Failed to send message'));
+      logging.error(err);
+      this.onShareCancelled();
+      return;
+    }
+    showToast(_('Message sent'));
+    this.onShareSuccess();
+  };
+
+  onShareCancelled = () => {
+    NavigationService.dispatch(navigateBack());
+  };
+
+  onShareSuccess = () => {
+    const { sendTo } = this.props;
+    switch (sendTo.type) {
+      case 'pm': {
+        const { selectedRecipients } = sendTo;
+        const { ownUserId } = this.props;
+        const recipients = pmKeyRecipientsFromIds(selectedRecipients, ownUserId);
+        const narrow = pmNarrowFromRecipients(recipients);
+        NavigationService.dispatch(replaceWithChat(narrow));
+        break;
+      }
+
+      case 'stream': {
+        const { stream } = sendTo;
+        const narrow = streamNarrow(stream);
+        NavigationService.dispatch(replaceWithChat(narrow));
+        break;
+      }
+
+      default:
+        ensureUnreachable(sendTo.type);
+        logging.error('Unknown type encountered in `sendTo`.');
+    }
```

Comment:
Cool. (Continues this thread: https://github.com/zulip/zulip-mobile/pull/4757#discussion_r670059993 .)

We do actually want the `onShareCancelled` call too, though. In general, because our types aren't guarantees, when we explicitly get into a case that should be impossible, we want to do *something* to bail out of the part of the code that's clearly confused. Throwing an exception is the default solution, but here `onShareCancelled` is handy and would be better.